python/ray/data/dataset.py

```diff
@@ -2602,30 +2602,12 @@ def schema(self, fetch_if_missing: bool = True) -> Optional["Schema"]:
         """
 
         # First check if the schema is already known from materialized blocks.
-        base_schema = self._plan.schema(fetch_if_missing=False)
+        base_schema = self._plan.schema(fetch_if_missing=fetch_if_missing)
         if base_schema is not None:
-            return Schema(base_schema)
-        if not fetch_if_missing:
-            return None
-
-        if self._plan.is_read_only():
-            # For read-only plans, there is special logic for fetching the
-            # schema from already known metadata
-            # (see `get_legacy_lazy_block_list_read_only()`). This requires
-            # the underlying logical plan to be read-only, so we skip appending
-            # the Limit[1] operation as we do in the else case below. There is
-            # no downside in this case, since it doesn't execute any read tasks.
-            base_schema = self._plan.schema(fetch_if_missing=fetch_if_missing)
+            schema = Schema(base_schema)
         else:
-            # Lazily execute only the first block to minimize computation.
-            # We achieve this by appending a Limit[1] operation to a copy
-            # of this Dataset, which we then execute to get its schema.
-            base_schema = self.limit(1)._plan.schema(fetch_if_missing=fetch_if_missing)
```

Comment:
I'm removing the `limit` optimization because I think we removed the limit pushdown rule, and I don't think there's any performance benefit in this code path (in both cases you would produce one output block).

Reply:
No, it'll only get one output and use that for the schema. Even if we add `limit(1)`, the behavior would be the same, although adding `limit` would introduce some complexity. https://github.com/ray-project/ray/blob/12a38fa5725b827f166f2be221ab8ac2d7118ee0/python/ray/data/_internal/plan.py#L377-L382

Reply:
wait, the `limit(1)` logic is not replicated in `plan.py` or anywhere else right? so we should still keep that logic?

Reply:
No, it's not replicated anywhere. I can re-add it